packages/modeling/src/operations/extrusions/earcut/assignHoles.js

```diff
@@ -0,0 +1,85 @@
+const { area } = require('../../../maths/utils')
+const { toOutlines } = require('../../../geometries/geom2')
+const { arePointsInside } = require('../../../geometries/poly2')
+
+/*
+ * Constructs a polygon hierarchy of solids and holes.
+ * The hierarchy is represented as a forest of trees. All trees shall be depth at most 2.
+ * If a solid exists inside the hole of another solid, it will be split out as its own root.
+ *
+ * @param {geom2} geometry
+ * @returns {Array} an array of polygons with associated holes
+ * @alias module:modeling/geometries/geom2.toTree
+ *
+ * @example
+ * const geometry = subtract(rectangle({size: [5, 5]}), rectangle({size: [3, 3]}))
+ * console.log(assignHoles(geometry))
+ * [{
+ *   "solid": [[-2.5,-2.5],[2.5,-2.5],[2.5,2.5],[-2.5,2.5]],
+ *   "holes": [[[-1.5,1.5],[1.5,1.5],[1.5,-1.5],[-1.5,-1.5]]]
+ * }]
```

Comment:
I think we need official support for this type of geometry, that is structured and gives more paths for optimizations. 
the code looks fine, I am just commenting this a as discussion point for the future.